design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+## Translation of errors
+
+Xlang's goal is to allow a component to be written using natural and familiar code for the language it is implemented in and for it to be consumed using natural and familiar code for the language which the consumer is written in.  To achieve this, xlang needs to translate errors from the natural form of one language to a natural form of that error in another language.
+
+But different languages have different errors which can occur and not all errors in one language always directly map to errors from another language.  In addition, maintaining a mapping of errors between each supported language is not scalable.  Lastly, there are a lot of errors which are domain specific and may not make sense in other languages and thereby not have equivalent error codes natively available.  Given this, xlang will support a limited set of common errors which can be used by the projection of a language to map to or from its error.  If the error being mapped has an equivalent in this set of common errors and the projection of the consumer's language provides an equivalent mapping from that mapped error, then that error will have a natural mapping in the other language.
+
+In the case there isn't an equivalent error to map to in xlang, then a generic error, xlang_fail, will be available to be used.  Note that in this case, the natural mapping of errors from one language to the other will be lost in the sense that it will not be an equivalent mapping.  In the case there isn't an equivalent error to map to in a consumer's language, it is left to the projection of that language to decide what error from its language to map to based on what makes sense for that language.  This can end up being a generic error.
+
+## Xlang errors
+
+Xlang will use an Int32, xlang_result, to represent its errors.  The representation of this 32-bit value will be based on the same model used in COM for HRESULTs to assist with sharing error codes, but will not necessarily be defined in this design note on a bit by bit basis as is in COM.  This is because xlang will not inherit all the errors used in COM and will instead define a set of common errors with their values defined.  This will allow xlang to avoid defining all the errors available in COM on each supported platform especially when most will not be used.  This will also allow for projections to avoid needing to maintain a large mapping of errors.
+
+The following errors will be defined in xlang:
+
+``` 
+xlang_fail (0x80004005) - To be used as a generic error when no equivalent or better error mapping is available
```

Comment:
I understand what you are saying.  During our last arch meeting, we discussed limiting the errors defined in xlang to ones which are likely to occur when projecting / translating a call because we want to discourage projected components from returning errors and rather have them using the try pattern.  In my next commit, I will remove a bunch of errors and scope it to ones that are likely to occur in projections.  The errors left I believe will mostly be ones that have posix mappings.  There are going to be errors which are not in the list of errors defined in xlang and those should map to xlang_fail.  We will support storing the posix error as part of the diagnostic informaton.

I don't believe we want to have a design where on Windows we use Windows errors and on Linux we use posix errors as that would most likely provide an inconsistent experience to consumers of xlang components when the errors get projected.

Reply:
We have decided to not base these errors directly off of HRESULT values and instead have our own values for them as it should be easy enough for projections for Window's APIs to have a switch which converts from one to the other.